### DISCUSSION TITLE
영속성 컨텍스트란? 영속성 컨텍스트의 이점은?

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#181] [Spring Data]
애플리케이션 & 데이터베이스 사이의 중간 계층 역할
1차 캐시 / 동일성 보장 / 쓰기 지연 / 변경 감지 / 지연 로딩 등등...

잘 모르는 부분이라 그런데 Hibernate만 제공하는 기능인가요? 아니면 JPA에서 정의한 기능인가요? 🥹

**Comment by J-I-H-O** [comment]
**<정의>**
영속성 컨텍스트는 애플리케이션과 데이터베이스 사이에서 객체를 보관하는 가상의 데이터베이스 같은 역할을 수행합니다.
엔티티 매니저를 통해 엔티티를 저장하거나 조회하면, 엔티티 매니저는 영속성 컨텍스트에 엔티티를 보관하고 관리합니다.
영속성 컨텍스트는 엔티티를 식별자 값으로 구분합니다. 따라서 영속 상태의 엔티티는 식별자 값이 존재해야 합니다.

**<영속성 컨텍스트의 이점>**

- **1차 캐시**
    영속성 컨텍스트 내부에는 영속 상태의 엔티티를 저장하는 1차 캐시가 존재합니다. `식별자 : 엔티티 객체` 의 key-value 쌍으로 저장됩니다. 1차 캐시가 존재함으로써 동일한 엔티티를 반복적으로 조회할 때 데이터베이스를 다시 조회하지 않고, 영속성 컨텍스트에 저장된 엔티티를 반환합니다.

    엔티티 조회 시 다음과 같은 과정을 거칩니다.
  ```
    1. 1차 캐시에 엔티티가 존재하는 경우, 해당 엔티티 객체 반환
    2. 1차 캐시에 엔티티가 존재하지 않는 경우, 데이터베이스에서 조회
    3. 조회한 데이터로 엔티티 객체를 생성해 1차 캐시에 저장 후 반환
  ```
- **영속 엔티티의 동일성 보장**
    영속성 컨텍스트가 관리하는 엔티티는 동일성을 보장합니다. 이는 같은 식별자로 조회한 엔티티는 같은 객체를 반환하기 때문입니다.
    > 동일성 비교 : `==` 비교
    
- **트랜잭션을 지원하는 쓰기 지연**
    엔티티 매니저는 내부 쿼리 저장소에 SQL을 모아두고, 트랜잭션을 커밋할 때 모아둔 쿼리를 DB로 보냅니다
    
- **변경 감지**
    JPA로 엔티티를 수정할 때에는 단순히 영속 상태의 엔티티의 값을 변경하면 됩니다. 이를 통해 개발자는 별도의 update 쿼리를 실행하지 않아도 됩니다.
    
    변경 감지는 다음과 같은 과정을 거칩니다.
    
  ```
    1. 엔티티와 스냅샷을 비교하여 변경된 엔티티를 찾음
    2. 변경된 엔티티가 있으면 update 쿼리를 생성하고, 쓰기 지연 저장소에 SQL을 저장
    3. 쓰기 지연 저장소의 SQL 을 플러시
    5. 데이터베이스 트랜잭션 커밋
  ```

**<플러시>** 
이때 플러시란, 영속성 컨텍스트의 변경 내용을 데이터베이스에 동기화하는 것을 뜻합니다. 
영속성 컨텍스트를 플러시 하는 방법은 다음과 같습니다.

- `em.flush()` 를 명시적으로 호출
- 트랜잭션 커밋 시 자동으로 호출
- JPQL 쿼리 실행 시 자동으로 호출

**Comment by le2sky** [comment]
영속성 컨텍스트는 엔티티를 영구 저장하는 환경입니다. 
엔티티 매니저로 엔티티를 저장하거나 조회하면 엔티티 매니저는 영속성 컨텍스트에 엔티티를 보관하고 관리합니다.
영속성 컨텍스트는 엔티티를 식별자 값으로 구분합니다. 또한, 트랜잭션을 커밋하는 순간 영속성 컨텍스트에 새로 저장된 엔티티를 데이터베이스에 반영합니다. (플러시)

영속성 컨텍스트의 이점은 다음과 같습니다.

- 1차 캐시 : 영속성 컨텍스트는 내부에 캐시를 가지고 있는데, 이를 1차 캐시라고 합니다.
- 동일성 보장 : em.find를 반복 호출해도 영속성 컨텍스트는 1차 캐시에 있는 같은 엔티티 인스턴스를 반환하기 때문에 동일성을 보장합니다.
- 트랜잭션을 지원하는 쓰기 지연 : 짤짤이 쿼리를 날려도 트랜잭션이 커밋되지 않으면 소용이 없기 때문에 지연 쓰기를 지원합니다.
- 변경 감지 : 스냅샷을 이용해 트랜잭션 커밋 시점에 변경사항을 데이터베이스에 반영합니다.
- 지연 로딩 : 필요한 경우에 연관 객체를 불러옵니다.

**Reply by le2sky** [reply]
추가로 영속성 컨텍스트의 지연쓰기나 변경감지에 대한 내용은 Synchronization to the Database 챕터에서 간접적으로 설명되는 것으로 보이네요.
